### ISSUE TITLE
get_feature_clusters generates ValueError when using mutual_information if feature df contains binary columns

### ISSUE THREAD

**Issue by BlackArbsCEO** [hudson-and-thames/mlfinlab#341]
**Describe the bug**
`get_feature_clusters` generates ValueError when using `mutual_information` for dependence_metric.

**To Reproduce**
```
X, y = ml.util.generate_dataset.get_classification_data(
    n_features=40,
    n_informative=5,
    n_redundant=30,
    n_samples=10000,
    sigma=0.1,
    random_state=87,
)

np.random.seed(87)
X["I_0"] = np.random.choice([0, 1], size=len(X))
X["I_1"] = np.random.choice([0, 1], size=len(X))

clusters = get_feature_clusters(
    X,
    dependence_metric="mutual_information",
    distance_metric=None,  # "angular",
    linkage_method=None,  # "average",
    n_clusters=None,
)

pprint(len(clusters))
pprint(clusters)
```

**Screenshots**
```
No feature/s found with low silhouette score. All features belongs to its respective clusters
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-260-c27b0f4e0b9e> in <module>
      4     distance_metric=None,  # "angular",
      5     linkage_method=None,  # "average",
----> 6     n_clusters=None,
      7 )
      8 

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/mlfinlab/clustering/feature_clusters.py in get_feature_clusters(X, dependence_metric, distance_metric, linkage_method, n_clusters, critical_threshold)
     53 
     54     if n_clusters is None and (distance_metric is None or linkage_method is None):
---> 55         return list(get_onc_clusters(dep_matrix.fillna(0))[1].values())  # Get optimal number of clusters
     56     if distance_metric is not None and (linkage_method is not None and n_clusters is None):
     57         n_clusters = len(get_onc_clusters(dep_matrix.fillna(0))[1])

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/mlfinlab/clustering/onc.py in get_onc_clusters(corr_mat, repeat)
    155     """
    156 
--> 157     return cluster_kmeans_top(corr_mat, repeat)

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/mlfinlab/clustering/onc.py in cluster_kmeans_top(corr_mat, repeat)
    115 
    116     max_num_clusters = corr_mat.shape[1] - 1
--> 117     corr1, clusters, silh = _cluster_kmeans_base(corr_mat, max_num_clusters=max_num_clusters, repeat=repeat)
    118 
    119     # Get cluster quality scores

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/mlfinlab/clustering/onc.py in _cluster_kmeans_base(corr_mat, max_num_clusters, repeat)
     63         for num_clusters in range(2, max_num_clusters + 1):
     64             kmeans_ = KMeans(n_clusters=num_clusters, n_jobs=1, n_init=1)
---> 65             kmeans_ = kmeans_.fit(distance)
     66             silh_ = silhouette_samples(distance, kmeans_.labels_)
     67             stat = (silh_.mean() / silh_.std(), silh.mean() / silh.std())

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/sklearn/cluster/_kmeans.py in fit(self, X, y, sample_weight)
    992                                 dtype=[np.float64, np.float32],
    993                                 order='C', copy=self.copy_x,
--> 994                                 accept_large_sparse=False)
    995         # verify that the number of samples given is larger than k
    996         if _num_samples(X) < self.n_clusters:

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/sklearn/base.py in _validate_data(self, X, y, reset, validate_separately, **check_params)
    418                     f"requires y to be passed, but the target y is None."
    419                 )
--> 420             X = check_array(X, **check_params)
    421             out = X
    422         else:

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/sklearn/utils/validation.py in inner_f(*args, **kwargs)
     71                           FutureWarning)
     72         kwargs.update({k: arg for k, arg in zip(sig.parameters, args)})
---> 73         return f(**kwargs)
     74     return inner_f
     75 

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/sklearn/utils/validation.py in check_array(array, accept_sparse, accept_large_sparse, dtype, order, copy, force_all_finite, ensure_2d, allow_nd, ensure_min_samples, ensure_min_features, estimator)
    644         if force_all_finite:
    645             _assert_all_finite(array,
--> 646                                allow_nan=force_all_finite == 'allow-nan')
    647 
    648     if ensure_min_samples > 0:

~/anaconda3/envs/mlfinlab/lib/python3.7/site-packages/sklearn/utils/validation.py in _assert_all_finite(X, allow_nan, msg_dtype)
     98                     msg_err.format
     99                     (type_err,
--> 100                      msg_dtype if msg_dtype is not None else X.dtype)
    101             )
    102     # for object dtype data, we only check for NaNs (GH-13254)

ValueError: Input contains NaN, infinity or a value too large for dtype('float64').
```

**Desktop (please complete the following information):**
```
mlfinlab 0.10.0

compiler   : GCC 7.3.0
system     : Linux
release    : 5.0.0-32-generic
machine    : x86_64
processor  : x86_64
CPU cores  : 12
```

**Comment by robolth** [comment]
I encountered the same bug in the same situation with both 'mutual_information' and 'information_variation' as a dependence_metric argument.

**Comment by PanPip** [comment]
@BlackArbsCEO thank you for reporting this issue.

I was able to reproduce it, it will be fixed in the soon future.

**Comment by PanPip** [comment]
This issue was caused by precision errors in the ONC module (the main diagonal of the correlation matrix was slightly higher than 1, so the distance matrix couldn't be calculated).

This issue is now fixed in PR #412
